atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/locks/Lockable.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.asts.locks;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.errorprone.annotations.CheckReturnValue;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockRequest;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.refreshable.Refreshable;
+import java.time.Duration;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+// See the LockableFactory comment for why keeping the value in the class ends up making things simpler from a local
+// lock management perspective.
+// Separately:
+// I, personally, am much more used to working with locks being attached to the data object. It's, imo, harder to screw
+// up and forget to lock / unlock things when a) you can't access the data without the lock and b) the error checker
+// will shout at you for not unlocking
+// TODO: Make the error checker shout at you
+// It also made the local isLocked check easier to manage rather than having a separate mapping (or, you don't have it
+// at all, but with so many threads and relatively high chance of some collisions, it seems like wasted RPCs otherwise)
+public final class Lockable<T> {
+    private static final SafeLogger log = SafeLoggerFactory.get(Lockable.class);
+    private final LockDescriptor lockDescriptor;
+    private final T inner;
+    private final TimelockService timeLock;
+    private final Refreshable<Duration> lockTimeout;
+
+    // Consider whether this should be a lock. It's simply an optimisation to avoid a check for timelock
+    // each time you want to test if it's locked by something locally, but we must make sure there are no false
+    // negatives. It's allowed that something local is working on this item and Timelock has released the lock due to
+    // expiry and this is still True (since that still lets us minimise conflicts by not working on it locally),
+    // but it should not be true if something local has stopped working on this item and closed the inner class
+    private final AtomicBoolean isLocked = new AtomicBoolean(false);
```

Comment:
so: basically, if this is locked then timelock _may_ have the lock token active, but if this is unlocked timelock _does not_ have the lock token active. Is that right?

Reply:
Looking at this more closely, this is actually _not_ right: if unlock fails because of a network blip...

Reply:
Discussed internally - nope! They can be out of sync, but isLocked must go to false even if Timelock calls fails (it's just an optimisation)